packages/grid/x-data-grid/src/hooks/features/selection/useGridSelection.ts

```diff
@@ -388,25 +392,67 @@ export const useGridSelection = (
         return;
       }
 
-      // Get the most recent params because the cell mode may have changed by another listener
-      const cellParams = apiRef.current.getCellParams(params.id, params.field);
-      const isEditMode = cellParams.cellMode === GridCellModes.Edit;
-      if (isEditMode) {
+      // Get the most recent cell mode because it may have been changed by another listener
+      if (apiRef.current.getCellMode(params.id, params.field) === GridCellModes.Edit) {
         return;
       }
 
       if (event.key === ' ' && event.shiftKey) {
         event.preventDefault();
-        handleSingleRowSelection(cellParams.id, event);
+        handleSingleRowSelection(params.id, event);
         return;
       }
 
       if (event.key.toLowerCase() === 'a' && (event.ctrlKey || event.metaKey)) {
         event.preventDefault();
         selectRows(apiRef.current.getAllRowIds(), true);
       }
+
+      if (isNavigationKey(event.key) && event.shiftKey) {
+        event.preventDefault();
+        const focusCell = gridFocusCellSelector(apiRef);
+        if (!focusCell || focusCell.id === params.id) {
+          return;
+        }
+
+        const isPreviousRowSelected = apiRef.current.isRowSelected(focusCell.id);
+        if (canHaveMultipleSelection) {
+          const newRowIndex = apiRef.current.getRowIndexRelativeToVisibleRows(focusCell.id);
+          const previousRowIndex = apiRef.current.getRowIndexRelativeToVisibleRows(params.id);
+
+          let start: number;
+          let end: number;
+
+          if (newRowIndex > previousRowIndex) {
+            if (isPreviousRowSelected) {
+              // We are navigating to the bottom of the page and adding selected rows
+              start = previousRowIndex;
+              end = newRowIndex - 1;
+            } else {
+              // We are navigating to the bottom of the page and removing selected rows
+              start = previousRowIndex;
+              end = newRowIndex;
+            }
+          } else if (isPreviousRowSelected) {
+            // We are navigating to the top of the page and removing selected rows
+            start = newRowIndex + 1;
+            end = previousRowIndex;
+          } else {
+            // We are navigating to the top of the page and adding selected rows
+            start = newRowIndex;
+            end = previousRowIndex - 1;
+          }
```

Comment:
Sounds easier to read width a 
```js
if (newRowIndex > previousRowIndex){
	if (isPreviousRowSelected) { ... }
	else { ... }
}
else {
	if (isPreviousRowSelected) { ... }
	else { ... }
}
````

such that we directly understand the first conditional level is about "moving top/bottom" second level is about adding/removing